Add actron_air climate integration

kclif9 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Contribute a new integration to Home Assistant - Actron Air Neo. This PR pertains to the Climate platform. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36728

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

kclif9 committed:
Add actronair neo climate integration

kclif9 committed:
Commit climate platform for actron air neo integration

joostlek reviewed (changes requested):
You marked rules complete in the quality scale.yaml which in fact are not complete. Please re review the rules and make sure the full quality scale is in there

kclif9 commented:
> You marked rules complete in the quality scale.yaml which in fact are not complete. Please re review the rules and make sure the full quality scale is in there

I believe I've implemented all the rules. Can you please advise which need to be fixed?

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    pairing_token = entry.data.get("pairing_token")
+    serial_number = entry.data.get("serial_number")
+
+    if not pairing_token or not serial_number:
+        raise ConfigEntryAuthFailed("Invalid authentication")
```
why would this be not there?

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    coordinator = ActronNeoDataUpdateCoordinator(hass, api, serial_number)
+    await coordinator.async_config_entry_first_refresh()
+
+    # Ensure coordinator data is not None
+    if coordinator.data is None:
+        raise ConfigEntryNotReady("Unable to retrieve data from the API")
```
I would expect the coordinator to raise UpdateFailed when that happens

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    entry.runtime_data = ActronAirNeoData(coordinator, api, ac_unit, serial_number)
+
+    # Store objects in hass.data
+    hass.data[DOMAIN][entry.entry_id] = {
+        "api": api,
+        "coordinator": coordinator,
+        "ac_unit": ac_unit,
+        "serial_number": serial_number,
+    }
```
Don't store these in hass.data

kclif9 replied:
I think these were stored but not actually referenced. Removed from code.

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    await async_setup_entry(hass, entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = True
+    for platform in PLATFORMS:
+        unload_ok = unload_ok and await hass.config_entries.async_forward_entry_unload(
+            entry, platform
+        )
+
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
```
This can be 1 call, please check other integrations

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    return unload_ok
+
+
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle migration of a config entry."""
+    if entry.version != 3:
+        new_data = {**entry.data}
+
+        if entry.version == 1:
+            if "pairing_token" not in new_data:
+                new_data["pairing_token"] = None
+        if entry.version == 2:
+            if "access_token" in new_data:
+                new_data["access_token"] = None
+
+        hass.config_entries.async_update_entry(entry, data=new_data, version=3)
+
+    return True
```
Users don't have this installed so a migration is not needed

kclif9 replied:
Removed and updated the config flow to V1.

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,451 @@
+        self.async_write_ha_state()
+
+
+class ActronZoneClimate(CoordinatorEntity, ClimateEntity):
```
Would it make sense to only include the system in the first PR so we can make the PR better scoped and faster to review?

kclif9 replied:
Will remove from this PR and add to a separate PR

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class ActronNeoConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class ActronNeoConfigFlow(ConfigFlow, domain=DOMAIN):
```

kclif9 replied:
Simplified to ConfigFlow throughout the file

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        self.api = None
+        self.ac_systems = None
+
+    async def async_step_user(self, user_input=None) -> config_entries.ConfigFlowResult:
```
```suggestion
    async def async_step_user(self, user_input=None) -> ConfigFlowResult:
```
Also please add typing to user_input

kclif9 replied:
Fixed

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+                errors=errors,
+            )
+
+        if not user_input.get("username") or not user_input.get("password"):
```
They are required, so this is not possible

kclif9 replied:
I think the referenced code was removed in the latest updates to config flow to add all systems and simplify the flow

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+
+        if user_input is None:
```
I recommend inverting this one to `if user_input is not None` as it will decrease the amount of `async_show_form` calls

kclif9 replied:
Fixed

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            )
+
+        try:
+            await self.api.request_pairing_token("HomeAssistant", "ha-instance-id")
```
what's this for

kclif9 replied:
I think the referenced code was removed in the latest updates to config flow to add all systems and simplify the flow

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        except ActronNeoAPIError:
+            errors["base"] = ERROR_API_ERROR
+
+        systems = await self.api.get_ac_systems()
+        self.ac_systems = systems.get("_embedded", {}).get("ac-system", [])
+        if not self.ac_systems:
+            errors["base"] = ERROR_NO_SYSTEMS_FOUND
```
What's this

kclif9 replied:
This is to retrieve the AC Systems from the relevant JSON path and error if none were found.

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+        if not self.ac_systems:
+            errors["base"] = ERROR_NO_SYSTEMS_FOUND
+
+        if len(self.ac_systems) > 1:
+            return self.async_show_form(
+                step_id="select_system",
+                data_schema=vol.Schema(
+                    {
+                        vol.Required("selected_system"): vol.In(
+                            {
+                                system["serial"]: system["description"]
+                                for system in self.ac_systems
+                            }
+                        )
+                    }
+                ),
+            )
```
Why don't we add all systems?

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,175 @@
+            },
+        )
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> config_entries.OptionsFlow:
+        """Get the options flow for this handler."""
+        return ActronNeoOptionsFlowHandler(config_entry)
+
+
+class ActronNeoOptionsFlowHandler(config_entries.OptionsFlow):
+    """Handle options for Actron Air Neo."""
+
+    def __init__(self, config_entry) -> None:
+        """Handle the initial setup."""
+        self.config_entry = config_entry
+
+    async def async_step_init(self, user_input=None) -> config_entries.ConfigFlowResult:
+        """Manage the options."""
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        return self.async_show_form(
+            step_id="init",
+            data_schema=vol.Schema({}),
+        )
```
This does nothing

joostlek commented on homeassistant/components/actronair_neo/const.py:
```diff
@@ -0,0 +1,9 @@
+"""Constants used by Actron Air Neo integration."""
+
+DOMAIN = "actronair_neo"
+PLATFORMS = ["climate"]
```
use `Platform` constant

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,154 @@
+class ActronNeoDataUpdateCoordinator(DataUpdateCoordinator):
+    """Custom coordinator for Actron Air Neo integration."""
+
+    def __init__(self, hass: HomeAssistant, api, serial_number) -> None:
```
typing

kclif9 replied:
Added typing

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,154 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class ActronNeoDataUpdateCoordinator(DataUpdateCoordinator):
```
```suggestion
class ActronNeoDataUpdateCoordinator(DataUpdateCoordinator[..]):
```
Please add the return type of `_async_update_data` here

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,154 @@
+
+    def _merge_incremental_update(self, full_state, incremental_data):
+        """Merge incremental updates into the full state."""
+        for key, value in incremental_data.items():
+            if key.startswith("@"):
+                continue
+
+            keys = key.split(".")
+            current = full_state
+
+            for part in keys[:-1]:
+                match = re.match(r"(.+)\[(\d+)\]$", part)
+                if match:
```
How does this work?

kclif9 replied:
The API provides an endpoint with updated data since the last full update. The incremental update code doesn't respond with neatly formatted JSON structure, rather similar to below. Hence this code to pull the structure apart and merge back into the full status update.

| data:
|   RemoteZoneInfo[3].LiveTemp_oC: 25.7

joostlek replied:
I think that that would be a concern to move to the library as this is device specific

kclif9 replied:
Done!

joostlek commented on homeassistant/components/actronair_neo/device.py:
```diff
@@ -0,0 +1,125 @@
+UNKNOWN_STATUS = "Unknown"
+
+
+class ACUnit:
```
I am not sure what this class is. Either this is a base entity (which should belong in entity.py) and be better typed, or this is some abstraction that should be abstracted away in the library

kclif9 replied:
Moved to climate.py and added typing

joostlek commented on homeassistant/components/actronair_neo/device.py:
```diff
@@ -0,0 +1,125 @@
+        return self._zone_number
+
+
+class ZonePeripheral:
```
I havent seen this been used

kclif9 replied:
Removed

joostlek commented on homeassistant/components/actronair_neo/manifest.json:
```diff
@@ -0,0 +1,17 @@
+  "name": "Actron Air Neo",
+  "codeowners": ["@kclif9"],
+  "config_flow": true,
+  "dependencies": [],
```
```suggestion
```

joostlek commented on homeassistant/components/actronair_neo/manifest.json:
```diff
@@ -0,0 +1,17 @@
+      "macaddress": "FC0FE7*"
+    }
+  ],
+  "documentation": "https://github.com/kclif9/actron-neo",
```
Please update this url

joostlek commented on homeassistant/components/actronair_neo/quality_scale.yaml:
```diff
@@ -0,0 +1,26 @@
+rules:
```
Please add all rules and assess them

kclif9 committed:
Merge branch 'home-assistant:dev' into add-actronairneo-climate-integration

kclif9 committed:
Not required as this is a new integration

kclif9 committed:
Simplified async_unload_entry call

kclif9 committed:
reload_entry not required

kclif9 commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload the Actron Air Neo integration."""
+    await async_unload_entry(hass, entry)
+    await async_setup_entry(hass, entry)
```
Removed reload_entry

kclif9 committed:
Simplify config_entries calls

kclif9 committed:
Simplified code and support multiple units

kclif9 committed:
Remove hass.data as this isn't referenced elsewhere

kclif9 committed:
Use ActronConfigEntry and remove ZoneClimate for a future PR

kclif9 committed:
Meet rule log_when_unavailable

kclif9 committed:
Merge branch 'home-assistant:dev' into add-actronairneo-climate-integration

kclif9 committed:
Remove dependencies and update documentation URL

kclif9 committed:
Remove unnecessary check

kclif9 committed:
Remove unnecessary check

kclif9 committed:
Use platform const

kclif9 committed:
Use platform const

kclif9 committed:
Simplify data schema

kclif9 committed:
Add typing

kclif9 commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    hass.data.setdefault(DOMAIN, {})
+
+    pairing_token = entry.data.get("pairing_token")
+    serial_number = entry.data.get("serial_number")
```
Done

kclif9 commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,88 @@
+from .models import ActronAirNeoData
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Done

kclif9 committed:
This is not used in this revision of the code

kclif9 committed:
Remove ACZone as is unused

kclif9 committed:
Add Any type

kclif9 committed:
Move ACUnit to climate.py file

kclif9 committed:
Move ACUnit to climate.py file

kclif9 committed:
Remove device.py file as is no longer required

kclif9 committed:
Add typing

kclif9 committed:
Remove import from device.py

kclif9 committed:
ACUnit moved to climate.py

kclif9 committed:
Merge branch 'home-assistant:dev' into add-actronairneo-climate-integration

kclif9 committed:
Removed circular reference to ac_unit

kclif9 committed:
Removed circular reference to ac_unit

kclif9 committed:
Removed circular reference to ac_unit

kclif9 committed:
Merge branch 'home-assistant:dev' into add-actronairneo-climate-integration

kclif9 committed:
Add missing quality scale items

kclif9 committed:
Update rules

kclif9 committed:
Add _async_update_data return type

kclif9 commented:
Thanks @joostlek for your review! I really appreciate the help. Hopefully that's all of the comments sorted!

kclif9 committed:
Remove setdomain per PR recommendation

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Remove redundant import

kclif9 commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,40 @@
+    coordinator = ActronNeoDataUpdateCoordinator(hass, api, serial_number)
+    await coordinator.async_config_entry_first_refresh()
+
+    entry.runtime_data = ActronAirNeoData(
+        pairing_token, coordinator, api, serial_number
+    )
```
Yeah, you're right on this one. It's totally redundant.

kclif9 committed:
Simplify coordinator and runtime_data

kclif9 commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,40 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ActronConfigEntry) -> bool:
+    """Set up Actron Air Neo integration from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
```
Removed domain from this section

kclif9 committed:
Import ActronConfigEntry from init

kclif9 commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,293 @@
+from .coordinator import ActronNeoDataUpdateCoordinator
+from .models import ActronAirNeoData
+
+type ActronConfigEntry = ConfigEntry[ActronAirNeoData]
```
Done

kclif9 committed:
Simplify unique_id

kclif9 committed:
Move system to the coordinator

kclif9 commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,293 @@
+    coordinator = instance.coordinator
+    serial_number = instance.serial_number
+
+    # Fetch system details and set up the AC Unit
+    system = await api.get_ac_systems()
```
Done

kclif9 committed:
Remove unnecessary code

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+                systems = await self.api.get_ac_systems()
+                self.ac_systems = systems.get("_embedded", {}).get("ac-system", [])
+
+                for system in self.ac_systems:
+                    serial_number = system["serial"]
+                    await self.async_set_unique_id(serial_number)
+                    self._abort_if_unique_id_configured()
+                    return self.async_create_entry(
+                        title=system["description"],
+                        data={
+                            CONF_API_TOKEN: self.api.pairing_token,
+                            CONF_DEVICE_ID: serial_number,
+                        },
```
This retrieves all of the systems retrieved by the API rather than before your previous review where it was just the user selected system serial_number on the config_entry.

joostlek replied:
Right, but we return in the for loop, so this will only create 1 entry. Can't we fetch all the entries during setup instead?

Like, instead of having 1 entry = 1 device, have 1 entry = 1 account = n devices

kclif9 replied:
Based on some of the other PR comments - changing this to be a username and will handle entity generation in the coordinator.

kclif9 commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,87 @@
+                    errors=errors,
+                )
+
+            if self.api:
```
I agree, however without that mypy won't let me commit the code in the event self.api is None.

homeassistant/components/actronair_neo/config_flow.py:60: error: "None" has no attribute "request_pairing_token"  [attr-defined]
homeassistant/components/actronair_neo/config_flow.py:61: error: "None" has no attribute "refresh_token"  [attr-defined]
homeassistant/components/actronair_neo/config_flow.py:65: error: "None" has no attribute "get_ac_systems"  [attr-defined]
homeassistant/components/actronair_neo/config_flow.py:68: error: "None" has no attribute "__iter__" (not iterable)  [attr-defined]
homeassistant/components/actronair_neo/config_flow.py:75: error: "None" has no attribute "pairing_token"  [attr-defined]

kclif9 replied:
Added assert to fix mypy & removed self.api false check

kclif9 replied:
added assert to resolve this

kclif9 committed:
Simplify climate objects

kclif9 commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,293 @@
+        self._serial_number = serial_number
+        self._ac_unit = ac_unit
+        self._attr_translation_placeholders = {"serial_number": self._serial_number}
+        self._attr_name = f"AC Unit {self._serial_number}"
```
I've simplified this. I'll work through translations as part of a separate PR.

joostlek replied:
My second question hasn't been answered, how does the device naming works?

kclif9 commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,293 @@
+    @property
+    def manufacturer(self) -> str:
+        """Return the manufacturer name."""
+        return self._manufacturer
```
Agreed. Simplified this mapping.

kclif9 committed:
assert to remove mypy error

kclif9 committed:
Shift the status merge logic into the API library

kclif9 committed:
Redundant code

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Add scan interval

kclif9 committed:
Utilise scan interval

kclif9 committed:
Add uniqueness to the API key when requesting a pairing token

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,36 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ActronConfigEntry) -> bool:
+    """Set up Actron Air Neo integration from a config entry."""
+
+    pairing_token = entry.data[CONF_API_TOKEN]
+    serial_number = entry.data[CONF_DEVICE_ID]
+
+    api = ActronNeoAPI(pairing_token=pairing_token)
+    await api.refresh_token()
```
Can refresh_token raise?

Maybe this can even be moved to the _async_setup function of the coordinator

kclif9 replied:
Moved to coordinator

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,36 @@
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
use the extended config entry

kclif9 replied:
Done

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,239 @@
+    async_add_entities(entities)
+
+
+class ActronSystemClimate(CoordinatorEntity, ClimateEntity):
+    """Representation of the Actron Air Neo system."""
+
+    _attr_has_entity_name = True
+    _attr_fan_modes = ["auto", "low", "medium", "high"]
+
+    def __init__(
+        self,
+        coordinator: ActronNeoDataUpdateCoordinator,
+    ) -> None:
+        """Initialize an Actron Air Neo unit."""
+        super().__init__(coordinator)
+        self._coordinator: ActronNeoDataUpdateCoordinator = coordinator
```
```suggestion
class ActronSystemClimate(CoordinatorEntity[ActronNeoDataUpdateCoordinator], ClimateEntity):
    """Representation of the Actron Air Neo system."""

    _attr_has_entity_name = True
    _attr_fan_modes = ["auto", "low", "medium", "high"]

    def __init__(
        self,
        coordinator: ActronNeoDataUpdateCoordinator,
    ) -> None:
        """Initialize an Actron Air Neo unit."""
        super().__init__(coordinator)
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,239 @@
+        self._model_name: str = self._status.get("AirconSystem", {}).get(
+            "MasterWCModel"
+        )
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return device information."""
+        return {
```
This can be set to `_attr_device_info` in the constructor instead

kclif9 replied:
Done

kclif9 committed:
Using extended config entry

kclif9 committed:
Moved logic to coordinator

kclif9 committed:
Adding Coordinator to class

kclif9 committed:
Moving to attr_device_info

kclif9 committed:
Changing the config_flow to save the pairing code only, and will handle serials in the coordinator

kclif9 committed:
Bump actron-neo-api

kclif9 committed:
Add support for adding multiple systems off a single configuration entry

kclif9 committed:
Simplify init

kclif9 committed:
Shift manufacturer to device_info

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Simplify const mapping

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,29 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ActronConfigEntry) -> bool:
+    """Set up Actron Air Neo integration from a config entry."""
+
+    # Initialize the data coordinator
+    pairing_token = entry.data[CONF_API_TOKEN]
+    coordinator = ActronNeoDataUpdateCoordinator(hass, pairing_token)
```
New thing happened, we should input the config entry to the coordinator

kclif9 replied:
What would be the result of this? From climate.py, I'm retrieving coordinator from the config entry. Wouldn't that be recursion?
From climate.py line 45
coordinator = entry.runtime_data

joostlek replied:
The result is that we are more explicit with what config entry it belongs to. I get why you call it recursion, but it's just a pointer in a pointer (I guess that's the best way I can explain it)

kclif9 replied:
Have added it

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        self,
+        coordinator: ActronNeoDataUpdateCoordinator,
+        serial_number: str,
+        description: str,
```
what's a description?

kclif9 replied:
Renamed this from description to name to better reflect what it's used for.

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        self._status = coordinator.data[self._serial_number]
+        self._name: str = description
+        self._attr_name: str = "AC Unit"
+        self._attr_unique_id: str = self._serial_number
```
```suggestion
        self._attr_unique_id: str = serial_number
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        self._serial_number: str = serial_number
+        self._status = coordinator.data[self._serial_number]
+        self._name: str = description
+        self._attr_name: str = "AC Unit"
```
AFAIK the climate entity would be the main feature of such device, so shouldn't we set the name to `_attr_name = None` so the entity name is the device name?

kclif9 replied:
Yeah that's a fair evaluation.

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+    ) -> None:
+        """Initialize an Actron Air Neo unit."""
+        super().__init__(coordinator)
+        self._coordinator: ActronNeoDataUpdateCoordinator = coordinator
```
```suggestion
```
This is already in `self.coordinator`

kclif9 replied:
Doesn't appear to actually be referenced in the code elsewhere so I've removed this reference.

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        self._name: str = description
+        self._attr_name: str = "AC Unit"
+        self._attr_unique_id: str = self._serial_number
+        self.attr_device_info = {
+            "identifiers": {(DOMAIN, self._serial_number)},
+            "name": self._name,
+            "manufacturer": "Actron Air",
+            "model": self._status.get("AirconSystem", {}).get("MasterWCModel"),
+            "sw_version": self._status.get("AirconSystem", {}).get(
+                "MasterWCFirmwareVersion"
+            ),
+            "serial_number": self._serial_number,
+        }
```
Let's use `DeviceInfo` object here

kclif9 replied:
Added DeviceInfo object

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+            "TemperatureSetpoint_Cool_oC"
+        )
+
+    @property
+    def supported_features(self) -> ClimateEntityFeature:
+        """Return supported features."""
+        return (
+            ClimateEntityFeature.TARGET_TEMPERATURE
+            | ClimateEntityFeature.FAN_MODE
+            | ClimateEntityFeature.TURN_ON
+            | ClimateEntityFeature.TURN_OFF
+        )
```
This can be set to a shorthand `_attr_supported_features`

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        fan_mode_without_cont = api_fan_mode.split("+")[0]
+        return FAN_MODE_MAPPING_REVERSE.get(fan_mode_without_cont, "AUTO")
+
+    @property
+    def fan_modes(self) -> list[str]:
+        """Return the list of available fan modes."""
+        return list(FAN_MODE_MAPPING.keys())
```
```suggestion
```

kclif9 replied:
Removed duplicated variable

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        """Return the list of available fan modes."""
+        return list(FAN_MODE_MAPPING.keys())
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the temperature unit."""
+        return UnitOfTemperature.CELSIUS
```
This can be set to a shorthand `_attr_temperature_unit`

kclif9 replied:
Fixed

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,229 @@
+        else:
+            raise ValueError(f"Mode {hvac_mode} is invalid.")
+        self._attr_target_temperature = temp
+        self.async_write_ha_state()
+
+    async def async_turn_on_continuous(self, continuous: bool) -> None:
+        """Set the continuous mode."""
+        await self._api.set_fan_mode(
+            self._serial_number, fan_mode=self._attr_fan_mode, continuous=continuous
+        )
+        self.async_write_ha_state()
```
instead of writing state, would it make more sense to just ask the coordinator to update?

kclif9 replied:
The problem is that the API can take 5-10 seconds to reflect the change after a request. I'm using a pre-emptive update to reflect the status change. Let me know if there are better ways of doing this.

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+        errors: dict[str, str] = {}
+
+        if user_input:
+            username = user_input["username"]
+            password = user_input["password"]
```
Let's use constants here

kclif9 replied:
Done

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.api = None
+
```
There's only one step, no need to have the `self.api` created in the constructor

kclif9 replied:
Fixed

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+
+            try:
+                instance_uuid = await instance_id.async_get(self.hass)
+                await self.api.request_pairing_token("HomeAssistant", instance_uuid)
```
why dow e do this?

kclif9 replied:
The API requires a unique ID to ensure that the instance on one device doesn't kill the access token for another instance.

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+            except ActronNeoAPIError:
+                errors["base"] = ERROR_API_ERROR
+
+            await self.async_set_unique_id(username.lower())
+            self._abort_if_unique_id_configured()
```
Can you change the username?

kclif9 replied:
Yeah, the username can be changed. Should this be switched to something else?

joostlek replied:
Ideally the unique id is strictly unique, so even if you rename the username we should be able to set up and recognize it's the same one. Do we get some token back for it maybe?

kclif9 replied:
Okay, I've replaced the username (email) in the config flow with the users GUID. I've then added a step to __init__ to update the email address

joostlek commented on homeassistant/components/actronair_neo/const.py:
```diff
@@ -0,0 +1,9 @@
+"""Constants used by Actron Air Neo integration."""
+
+DOMAIN = "actronair_neo"
+PLATFORM = ["climate"]
```
Please use the `Platform` enum

kclif9 replied:
Updated

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,43 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=15)
```
15 seconds is quite low, are we sure we want that?

kclif9 replied:
Updated to 30 sec.

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,43 @@
+SCAN_INTERVAL = timedelta(seconds=15)
+
+
+class ActronNeoDataUpdateCoordinator(DataUpdateCoordinator[dict]):
```
can we type the dict more specific?

kclif9 replied:
Added

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,43 @@
+
+    async def _async_update_data(self) -> dict:
+        """Fetch updates and merge incremental changes into the full state."""
+        if self.api.access_token is None:
+            await self.api.refresh_token()
+
+        if self.systems is None:
+            self.systems = await self.api.get_ac_systems()
```
Move these 2 actions to `_async_setup`

kclif9 committed:
Rename description variable to name to better reflect what the attribute is used for

kclif9 committed:
Remove unused variable

kclif9 committed:
Set to none as climate is the main feature of the device.

kclif9 committed:
Simplify the serial_number variable

kclif9 committed:
Set to short hand

kclif9 committed:
Shorthand _attr_supported_features

kclif9 committed:
Shorthand _attr_supported_features

kclif9 committed:
Update frequency to 30 sec

kclif9 committed:
Added DeviceInfo Object

kclif9 committed:
This is already in _attr_fan_modes

kclif9 committed:
Fix invalid variable

kclif9 committed:
Line is redundant

kclif9 committed:
Bump actron-neo-api to 0.1.45

kclif9 committed:
Bump actron-neo-api to 0.1.46

kclif9 committed:
Fix status change reversion

kclif9 committed:
simplify api and use constants

kclif9 committed:
use platform const

kclif9 committed:
fix typing and add async_setup

kclif9 committed:
Add zone climate sensors

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,353 @@
+
+        zones = coordinator.data[serial_number].get("RemoteZoneInfo", [])
+
+        zone_map = dict(enumerate(zones, start=0))
```
So the zone number is based on the order in the dict. Does that ever change?

kclif9 replied:
No, the zone number is always in the same order, as this is presented based on the zone order on the zone controller.

joostlek replied:
Do we have something more unique for the zone identifier?

Could the possibility that this changes if someone changes their set up?

kclif9 replied:
There's nothing else to reference aside from the zone name which is not uniquely enforced. The zone identifier is based on the physical port on the zone controller, so is unlikely to be changed.

kclif9 committed:
Use user id for entry unique id

kclif9 committed:
Update email on __init__

kclif9 committed:
Adding config entry to coordinator

kclif9 committed:
Refactor Actron Air Neo integration: enhance coordinator, add diagnostics and repair flows, update requirements

kclif9 committed:
Revert removal of user_input in config flow

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Update quality scale: change test coverage and translation statuses to 'todo'

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        | ClimateEntityFeature.TURN_ON
+        | ClimateEntityFeature.TURN_OFF
+    )
+    _attr_entity_registry_enabled_default = True
```
```suggestion
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        super().__init__(coordinator)
+        self._serial_number: str = serial_number
+        self._name: str = name
+        self._attr_name: None = None
```
`_attr_name` can be set outside of the constructor

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        self._serial_number: str = serial_number
+        self._name: str = name
+        self._attr_name: None = None
+        self._attr_unique_id: str = self._serial_number
```
```suggestion
        self._attr_unique_id = serial_number
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        self._name: str = name
+        self._attr_name: None = None
+        self._attr_unique_id: str = self._serial_number
+        initial_status = coordinator.get_status(serial_number)
```
instead of using `initial_status`, just use `self._status`

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        self._attr_unique_id: str = self._serial_number
+        initial_status = coordinator.get_status(serial_number)
+        self._attr_device_info: DeviceInfo = DeviceInfo(
+            identifiers={(DOMAIN, self._serial_number)},
+            name=initial_status.ac_system.system_name,
+            manufacturer="Actron Air",
+            model=initial_status.ac_system.master_wc_model,
+            sw_version=initial_status.ac_system.master_wc_firmware_version,
+            serial_number=self._serial_number,
```
```suggestion
            identifiers={(DOMAIN, serial_number)},
            name=initial_status.ac_system.system_name,
            manufacturer="Actron Air",
            model=initial_status.ac_system.master_wc_model,
            sw_version=initial_status.ac_system.master_wc_firmware_version,
            serial_number=serial_number,
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+    @property
+    def hvac_mode(self) -> HVACMode:
+        """Return the current HVAC mode."""
+        system_state = self._status.user_aircon_settings.is_on
+        if not system_state:
```
```suggestion
        if not self._status.user_aircon_settings.is_on:
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        mode = self._status.user_aircon_settings.mode
+        return HVAC_MODE_MAPPING.get(mode, HVACMode.OFF)
+
+    @property
+    def hvac_modes(self) -> list[HVACMode]:
+        """Return HVAC Modes."""
+        return [HVACMode.OFF, HVACMode.COOL, HVACMode.HEAT, HVACMode.AUTO]
```
this can be set to `_attr_hvac_modes = [..]`

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        | ClimateEntityFeature.TURN_ON
+        | ClimateEntityFeature.TURN_OFF
+    )
+    _attr_entity_registry_enabled_default = True
```
```suggestion
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+    ) -> None:
+        """Initialize an Actron Air Neo unit."""
+        super().__init__(coordinator)
+        self._serial_number: str = serial_number
```
```suggestion
        self._serial_number = serial_number
```

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+        self._zone_id: int = zone.zone_id
+        self._attr_name: None = None
+        self._attr_unique_id: str = f"{serial_number}_zone_{zone.zone_id}"
+        self._attr_device_info: DeviceInfo = DeviceInfo(
+            identifiers={(DOMAIN, self._attr_unique_id)},
+            name=zone.title,
+            manufacturer="Actron Air",
+            model="Zone",
+            suggested_area=zone.title,
+        )
```
use `via_device` to link back to the system device

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,125 @@
+        if serial_number in self.status_objects:
+            return self.status_objects[serial_number]
+
+        # Fallback to getting from the API state manager (shouldn't be needed in normal operation)
+        return self.api.state_manager.get_status(serial_number)
```
Does this do an API request and thus a blocking IO operation?

kclif9 replied:
Yeah that's right

joostlek replied:
IO operations must be executed in an executor job, and thus in an async context

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,263 @@
+    @property
+    def _zone(self) -> ActronAirNeoZone:
+        """Get the current zone data from the coordinator."""
+        status = self.coordinator.get_status(self._serial_number)
+        for zone in status.remote_zone_info:
+            if zone.zone_id == self._zone_id:
+                return zone
+        return None
```
I would strongly recommend to update the library to return the zones as `dict[str, ActronAirNeoZone]`. This way you don't have to do a loop every time but just once when creating the dict

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+            errors=errors,
+        )
+
+    async def async_step_reauth(
+        self, entry_data: Mapping[str, Any]
+    ) -> ConfigFlowResult:
+        """Handle reauthorization request."""
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, str] | None = None
+    ) -> ConfigFlowResult:
+        """Handle reauthorization confirmation."""
+        return await self.async_step_user()
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Reconfigure the integration."""
+        return await self.async_step_user()
```
Let's keep reauth and reconfigure out for now as that adds a ton of tests

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,96 @@
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+            _LOGGER.debug("Connecting to Actron Neo API")
+            try:
+                api = ActronNeoAPI(username, password)
+            except ActronNeoAuthError:
+                errors["base"] = ERROR_INVALID_AUTH
+                return self.async_show_form(
+                    step_id="user",
+                    data_schema=ACTRON_AIR_SCHEMA,
+                    errors=errors,
+                )
+
+            try:
+                instance_uuid = await instance_id.async_get(self.hass)
+                await api.request_pairing_token("HomeAssistant", instance_uuid)
+                await api.refresh_token()
+            except ActronNeoAPIError:
+                errors["base"] = ERROR_API_ERROR
+                return self.async_show_form(
+                    step_id="user",
+                    data_schema=ACTRON_AIR_SCHEMA,
+                    errors=errors,
+                )
```
I would combine these 2 try blocks and move line 63-72 to the `else` block, so then you can just re-use the `self.async_show_form` from line 74

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,125 @@
+        self, hass: HomeAssistant, entry: ActronConfigEntry, pairing_token: str
+    ) -> None:
+        """Initialize the coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="Actron Neo Status",
+            update_interval=SCAN_INTERVAL,
+        )
+        self.api = ActronNeoAPI(pairing_token=pairing_token)
+        self.entry = entry
```
instead, pass the entry to the constructor of the coordinator

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,125 @@
+        self.last_update_success = False
+        self.last_seen: dict[str, datetime] = {}
+        self.auth_error_count = 0
+        self.hass = hass
```
```suggestion
```

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,125 @@
+                # Store the actual status object
+                self.status_objects[serial] = status
+                # Also maintain the serialized data for the DataUpdateCoordinator
+                status_data[serial] = (
+                    status.to_dict() if hasattr(status, "to_dict") else vars(status)
+                )
```
why would it not have to_dict?

kclif9 replied:
It should have just been vars(status) and not to_dict

joostlek commented on homeassistant/components/actronair_neo/diagnostics.py:
Please remove diagnostics for a follow up PR

joostlek commented on homeassistant/components/actronair_neo/repairs.py:
we already have reauth for this, so we don't need a reimplementation for the repair so this file can be removed :)

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Remove default entity registry enabled attribute from ActronSystemClimate class

kclif9 committed:
Refactor ActronSystemClimate and ActronZoneClimate classes: streamline unique ID assignment, update HVAC modes, and enhance device info initialization.

kclif9 committed:
Refactor ActronNeoConfigFlow: remove unused reauthorization and reconfiguration steps; update quality scale to reflect reauthentication and reconfiguration flows as todo

kclif9 committed:
Remove diagnostics and repairs support for Actron Air Neo integration; update quality scale to reflect diagnostics and repair issues as todo

kclif9 committed:
Remove unused 'hass' attribute from ActronNeoDataUpdateCoordinator class

kclif9 committed:
Remove stale authentication issue registration from ActronNeoDataUpdateCoordinator

kclif9 committed:
Simplify status data serialization in ActronNeoDataUpdateCoordinator

kclif9 committed:
Refactor ActronNeoDataUpdateCoordinator: remove redundant entry assignment in constructor

kclif9 committed:
Refactor ActronNeoDataUpdateCoordinator: streamline status update logic and remove unused status data serialization

kclif9 committed:
Refactor ActronSystemClimate: use status attribute for device info initialization

kclif9 committed:
Update actron-neo-api requirement to version 0.1.76 in manifest.json

kclif9 committed:
Refactor ActronZoneClimate: simplify zone data retrieval logic

kclif9 committed:
Refactor ActronNeoConfigFlow: streamline error handling and improve code readability

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,253 @@
+            identifiers={(DOMAIN, serial_number)},
+            name=self._status.ac_system.system_name,
+            manufacturer="Actron Air",
+            model=self._status.ac_system.master_wc_model,
```
What is this model? Can you give an example?

kclif9 replied:
HA shows the model as NTW-1000 which is the wall controller

joostlek replied:
Is that a neverchangeable model? Isn't that more fit to be the model_id?

kclif9 replied:
Yeah that's true. Updated

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,253 @@
+        super().__init__(coordinator)
+        self._serial_number = serial_number
+        self._zone_id: int = zone.zone_id
+        self._attr_name: None = None
```
attr name can be set outside of the constructor

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,253 @@
+        await self._status.user_aircon_settings.set_continuous_mode(enabled=continuous)
+
+
+class ActronZoneClimate(
+    CoordinatorEntity[ActronNeoDataUpdateCoordinator], ClimateEntity
+):
```
I think you can extract a base Climate entity and deduplicate a few things you set like serial number, name, has_entity_name

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
```
```suggestion
```

kclif9 replied:
Yeah that's kinda dumb

joostlek commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,64 @@
+                await api.request_pairing_token("HomeAssistant", instance_uuid)
+                await api.refresh_token()
+                user_data = await api.get_user()
+                await self.async_set_unique_id(user_data["id"])
+
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=username,
+                    data={
+                        CONF_API_TOKEN: api.pairing_token,
+                    },
+                )
+            except ActronNeoAuthError:
+                errors["base"] = ERROR_INVALID_AUTH
+            except ActronNeoAPIError:
+                errors["base"] = ERROR_API_ERROR
```
```suggestion
            except ActronNeoAuthError:
                errors["base"] = ERROR_INVALID_AUTH
            except ActronNeoAPIError:
                errors["base"] = ERROR_API_ERROR
            else:
                await self.async_set_unique_id(user_data["id"])
                self._abort_if_unique_id_configured()
                return self.async_create_entry(
                    title=username,
                    data={
                        CONF_API_TOKEN: api.pairing_token,
                    },
                )
```

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+            config_entry=entry,
+        )
+        self.api = ActronNeoAPI(pairing_token=pairing_token)
+        self.last_update_success = False
```
I believe the last_update_success is a variable by the base coordinator

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+                err,
+            )
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+        else:
+            return self.status_objects[serial]
```
why do we only return one status objects?

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,109 @@
+            _LOGGER.error("API error while setting up Actron Neo integration: %s", err)
+            raise
+
+    async def _async_update_data(self) -> dict[str, Any]:
+        """Fetch updates and merge incremental changes into the full state."""
+        try:
+            await self.api.update_status()
+            self.last_update_success = True
+            self.auth_error_count = 0
+            current_time = dt_util.utcnow()
+
+            for system in self.systems:
+                serial = system.get("serial")
+                self.last_seen[serial] = current_time
+                status = self.api.state_manager.get_status(serial)
+                self.status_objects[serial] = status
```
Wouldn't it make more sense to create a coordinator for every device? This way you can simplify the logic, and then if a single device fails, it doesn't render all devices unavailable at once.

kclif9 replied:
Agreed - Updated the logic. Hopefully that makes sense.

kclif9 committed:
Refactor ActronSystemClimate and ActronZoneClimate: remove type annotations for name attributes

kclif9 committed:
Refactor ActronNeoConfigFlow: remove unnecessary constructor and streamline error handling

kclif9 committed:
Refactor ActronSystemClimate: correct fan mode mapping for consistency

kclif9 committed:
Refactor ActronSystemClimate: use mappings for fan and HVAC modes for consistency

kclif9 committed:
Refactor climate entities: extract common functionality into BaseClimateEntity and streamline ActronSystemClimate and ActronZoneClimate classes

kclif9 committed:
Refactor Actron Air Neo integration: streamline coordinators and update data handling for improved performance and clarity

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek commented on homeassistant/components/actronair_neo/__init__.py:
```diff
@@ -0,0 +1,40 @@
+        await coordinator.async_config_entry_first_refresh()
+        system_coordinators[system["serial"]] = coordinator
+
+    user_data = await api_coordinator.api.get_user()
+    hass.config_entries.async_update_entry(entry, title=user_data["email"])
```
the title of a config entry is userland, so we should not change it

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,258 @@
+        self._serial_number = coordinator.serial_number
+        self._name = name
+
+    @property
+    def min_temp(self) -> float:
+        """Return the minimum temperature that can be set."""
+        return self._this_device().min_temp
+
+    @property
+    def max_temp(self) -> float:
+        """Return the maximum temperature that can be set."""
+        return self._this_device().max_temp
+
+    def _this_device(self):
+        """Get the device this entity represents (to be implemented by subclasses)."""
+        raise NotImplementedError
```
I would not put this one in the shared class

joostlek commented on homeassistant/components/actronair_neo/climate.py:
```diff
@@ -0,0 +1,258 @@
+        super().__init__(coordinator, zone.title)
+        serial_number = coordinator.serial_number
+        self._zone_id: int = zone.zone_id
+        self._attr_unique_id: str = f"{serial_number}_zone_{zone.zone_id}"
```
```suggestion
        self._attr_unique_id: str = f"{self._serial_number}_zone_{zone.zone_id}"
```

joostlek commented on homeassistant/components/actronair_neo/coordinator.py:
```diff
@@ -0,0 +1,100 @@
+PARALLEL_UPDATES = 0
+
+
+class ActronNeoApiCoordinator:
+    """Coordinator for Actron Neo API."""
+
+    def __init__(self, hass: HomeAssistant, entry: ActronNeoConfigEntry) -> None:
+        """Initialize the coordinator."""
+        self.hass = hass
+        self.entry = entry
+        self.api = ActronNeoAPI(pairing_token=entry.data[CONF_API_TOKEN])
+        self.systems: list[ActronAirNeoACSystem] = []
+
+    async def async_setup(self) -> bool:
+        """Perform initial setup, including refreshing the token."""
+        try:
+            await self.api.refresh_token()
+            systems = await self.api.get_ac_systems()
+            self.systems = systems
+        except ActronNeoAuthError:
+            _LOGGER.error(
+                "Authentication error while setting up Actron Neo integration"
+            )
+            raise
+        except ActronNeoAPIError as err:
+            _LOGGER.error("API error while setting up Actron Neo integration: %s", err)
+            raise
+        return True
```
Why do we have this separate class? Can we merge it with the coordinator?

kclif9 replied:
I separated it into its own class as the same API endpoint can return multiple systems. It made sense to me to break it into two separate classes such that the initial step in initalisation connects to the API and retrieves systems, then the second step creates the classes for the system coordinators. 

This way there cannot be any confusion between having multiple API endpoints (such as if you have two Actron Neo accounts) and multiple systems

joostlek replied:
How do you mean the second step? currently I kinda see this as a wrapper for the client, and a part of a data holder, which is a role that the coordinator already has

kclif9 replied:
In the init file, I connect to the API to pull in the systems that are connected to the account, then have a separate coordinator instance for each system. This significantly reduces the complexity of the integration. 

My concern behind merging them back together, and together with your previous recommendations around simplifying the coordinator, it may get complicated if I'm instantiating a coordinator, then creating additional instances for multiple systems. Separating the logic for the two makes sense to me as they do serve separate purposes, and makes it easier when you have multiple systems under a single account. 

To make the naming clearer, I've renamed the ApiCoordinator to ApiClient as that reflects more accurately what its purpose is. Are you okay with this approach?

kclif9 committed:
Refactor Actron Air Neo climate entities: remove unused methods and streamline temperature property access

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Refactor Actron Air Neo integration: rename API coordinator to API client for clarity

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

zweckj commented on homeassistant/components/actronair_neo/quality_scale.yaml:
```diff
@@ -0,0 +1,70 @@
+  docs-troubleshooting: done
+  docs-use-cases: done
+  dynamic-devices: done
+  entity-category: done
```
exempt

zweckj commented on homeassistant/components/actronair_neo/quality_scale.yaml:
```diff
@@ -0,0 +1,70 @@
+  exception-translations: todo
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
```
exempt, nothing to repair yet

zweckj commented on homeassistant/components/actronair_neo/quality_scale.yaml:
```diff
@@ -0,0 +1,70 @@
+  icon-translations: todo
+  reconfiguration-flow: todo
+  repair-issues: todo
+  stale-devices: done
```
```suggestion
  stale-devices: todo
```

zweckj commented on tests/components/actronair_neo/const.py:
```diff
@@ -0,0 +1,9 @@
+"""Constants used by Actron Air Neo integration."""
+
+DOMAIN = "actronair_neo"
+PLATFORMS = ["climate", "sensor", "switch"]
```
```suggestion
PLATFORMS = ["climate"]
```

zweckj commented on tests/components/actronair_neo/const.py:
```diff
@@ -0,0 +1,9 @@
+"""Constants used by Actron Air Neo integration."""
+
+DOMAIN = "actronair_neo"
```
import that from components

zweckj commented on tests/components/actronair_neo/const.py:
```diff
@@ -0,0 +1,9 @@
+DOMAIN = "actronair_neo"
+PLATFORMS = ["climate", "sensor", "switch"]
+
+ERROR_API_ERROR = "api_error"
+ERROR_INVALID_AUTH = "invalid_auth"
+ERROR_NO_SYSTEMS_FOUND = "no_systems_found"
+ERROR_UNKNOWN = "unknown_error"
```
```suggestion
```
rather hardcode in code

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+        yield mock_api
+
+
+async def test_user_flow_single_system(hass: HomeAssistant, mock_actron_api) -> None:
```
missing type

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    assert result["data"] == {
+        "pairing_token": "test_pairing_token",
+        "serial_number": "12345",
+    }
```
assert unique id

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    assert result["type"] == "create_entry"
+    assert result["title"] == "Living Room"
+    assert result["data"] == {
+        "pairing_token": "test_pairing_token",
+        "serial_number": "12345",
```
use the CONFs here

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    assert result["step_id"] == "user"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], {"username": "test_user", "password": "test_pass"}
```
same here (you can import them from components) and everywhere else in that test

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+
+# Fixtures for the mock ActronNeoAPI
+@pytest.fixture
+def mock_actron_api():
```
return type

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+        "custom_components.actron_neo.config_flow.ActronNeoAPI", autospec=True
+    ) as mock_api:
+        yield mock_api
+
```
please add a `mock_setup_entry` fixture to avoid actually setting up the integration in the config flow tests

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+    mock_api_instance.request_pairing_token.side_effect = ActronNeoAuthError
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```
import that please

zweckj commented on tests/components/actronair_neo/test_config_flow.py:
```diff
@@ -0,0 +1,155 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == "form"
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```
everywhere with the enum and `is` comparison

zweckj reviewed (dismissed):
also, CI is failing with some import errors from the library

kclif9 committed:
Refactor fan and HVAC mode mappings for Actron Air Neo integration

kclif9 committed:
removing default version

kclif9 committed:
Fix condition check for user input in Actron Neo config flow

kclif9 commented on homeassistant/components/actronair_neo/config_flow.py:
```diff
@@ -0,0 +1,61 @@
+            username = user_input[CONF_USERNAME]
+            password = user_input[CONF_PASSWORD]
+            _LOGGER.debug("Connecting to Actron Neo API")
+            try:
+                api = ActronNeoAPI(username, password)
```
The connection to the api could fail if the credentials are invalid, hence why I kept that in there on a previous comment on this PR

kclif9 committed:
Refactor error handling in Actron Neo config flow by replacing constants with string literals

kclif9 committed:
Refactor Actron Air Neo integration by reorganizing imports and removing unused constants

kclif9 committed:
Refactor Actron Air Neo integration by removing const.py and updating logger imports

kclif9 committed:
Refactor Actron Air Neo integration by removing const.py and updating test configuration flow

kclif9 committed:
Update quality scale for Actron Air Neo integration by changing action-exceptions to todo and adding status comments for various parameters

kclif9 committed:
fix circular references

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

JagadishDhanamjayam commented:
> > FYI there is also a PR from actronair directly #142369. I think this one is already in a better shape though, so maybe you can join forces and get official buy-in from them.
> 
> @zweckj I've had a review of the code in the other PR and agree that we should proceed with this PR. The API code not being in its own module published to PyPi will take significant effort to migrate - Perhaps @JagadishDhanamjayam can contribute to the actronair_neo PyPi library to migrate the OAUTH flow to it, and this can be pushed as part of a separate PR.
> 
> For now, I feel this integration should be ready to go into production, so long as @zweckj is okay with the changes made.

@kclif9 @zweckj, 
I am not sure what do you mean by API code not being in its own module?
API code is in its own module in PyPi, here are the links to it
https://github.com/Actron-Air/ActronAirAPI
https://pypi.org/project/actronair-api/

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

JagadishDhanamjayam commented:
> I think the approach outlined by @kclif9 sounds reasonable, @JagadishDhanamjayam is that something we can agree on?

Thanks @zweckj and @kclif9 – I fully agree with the proposed approach and appreciate the thoughtful path forward. I’m keen to collaborate closely to ensure long-term support for Actron integrations within Home Assistant.

In summary:

1. I’ll be added as a code owner (representing ActronAir).

2. @kclif9 will rename the integration from actronair_neo to actronair, broadening the scope to all ActronAir products.

3. Either I can contribute the OAuth device code flow logic from my [existing PR](https://github.com/home-assistant/core/pull/142369), or @kclif9 can integrate it directly.

4. I will review and work towards consolidating the current PyPI library with the broader needs

5. Please also consider incorporating the brand assets from our [brands PR](https://github.com/home-assistant/brands/pull/6857) for consistency with ActronAir branding.

Looking forward to working together to deliver the best experience for Home Assistant users.

joostlek commented:
If you already have a PR open for the brands, I think it's fine to reuse that one!

kclif9 committed:
Update Actron Air Neo integration with OAuth2 flow and logging enhancements

kclif9 committed:
Rename integration directory

kclif9 committed:
Rename integration

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 commented:
Hello, 

Added @JagadishDhanamjayam to the integration code owners. 
Renamed directory to actronneo to reflect the direction & use the branding from the PR @JagadishDhanamjayam submitted for brands. 
The config flow has been updated to oAuth2.

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek commented on homeassistant/components/actronair/config_flow.py:
```diff
@@ -0,0 +1,74 @@
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+
+        if self._api is not None and self._device_code is not None:
+            try:
+                token_data = await self._api.poll_for_token(self._device_code)
+            except ActronNeoAuthError as err:
+                _LOGGER.error("Error checking authorization: %s", err)
+                errors["base"] = "oauth2_error"
+
+            if token_data is None:
+                errors["base"] = "authorization_pending"
+            else:
+                user_data = await self._api.get_user_info()
+                await self.async_set_unique_id(user_data["id"])
+                self._abort_if_unique_id_configured()
+                return self.async_create_entry(
+                    title=user_data["email"],
+                    data={
+                        CONF_API_TOKEN: self._api.refresh_token_value,
+                    },
+                )
+
+        else:
+            try:
+                self._api = ActronNeoAPI()
+                device_code_response = await self._api.request_device_code()
+
+            except ActronNeoAuthError as err:
+                _LOGGER.error("OAuth2 flow failed: %s", err)
+                return self.async_abort(reason="oauth2_error")
+
+            self._device_code = device_code_response["device_code"]
+            self._user_code = device_code_response["user_code"]
+            self._verification_uri = device_code_response["verification_uri_complete"]
+            self._expires_minutes = str(device_code_response["expires_in"] // 60)
+
+        return self.async_show_form(
+            step_id="user",
```
Looking at the strings, this looks like device auth, I would recommend to look at the tado integration or the github integration

joostlek commented on homeassistant/components/actronair/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+class ActronNeoRuntimeData:
+    """Runtime data for the Actron Air Neo integration."""
+
+    api_client: "ActronNeoApiClient"
+    system_coordinators: dict[str, "ActronNeoSystemCoordinator"]
```
no need to add this in quotes

kclif9 replied:
Both Ruff & Pylace error out if they're not in quotes as they're defined lower in the coordinator.

joostlek commented on homeassistant/components/actronair/coordinator.py:
```diff
@@ -0,0 +1,103 @@
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class ActronNeoApiClient:
+    """Client for Actron Neo API."""
+
+    def __init__(self, hass: HomeAssistant, entry: ActronNeoConfigEntry) -> None:
+        """Initialize the client."""
+        self.hass = hass
+        self.entry = entry
+        self.api = ActronNeoAPI(refresh_token=entry.data[CONF_API_TOKEN])
+        self.systems: list[ActronAirNeoACSystem] = []
+
+    async def async_setup(self) -> bool:
+        """Perform initial setup, including refreshing the token."""
+        try:
+            self.systems = await self.api.get_ac_systems()
+            await self.api.update_status()
+        except ActronNeoAuthError:
+            _LOGGER.error(
+                "Authentication error while setting up Actron Neo integration"
+            )
+            raise
+        except ActronNeoAPIError as err:
+            _LOGGER.error("API error while setting up Actron Neo integration: %s", err)
+            raise
+        return True
```
Okay so this is just an object that fetches info and has no other tasks. In the future, this might become its own coordinator. I think we can just move this code to async setup entry in `__init__.py`, as there is no real need for this to be an object

kclif9 committed:
map fan modes to const

kclif9 committed:
remove unused var

kclif9 committed:
return None instead of OFF

kclif9 commented on homeassistant/components/actronair/climate.py:
```diff
@@ -0,0 +1,260 @@
+        temp = kwargs.get("temperature")
+        await self._status.user_aircon_settings.set_temperature(temperature=temp)
+
+    async def async_turn_on_continuous(self, continuous: bool) -> None:
+        """Set the continuous mode."""
+        await self._status.user_aircon_settings.set_continuous_mode(enabled=continuous)
```
Yep, good pickup. That should be in switch (later PR)

kclif9 committed:
remove unused funct

kclif9 committed:
remove .get()

kclif9 committed:
return None instead of OFF

kclif9 committed:
use ATTR_TEMPERATURE

kclif9 committed:
fix type annotation

kclif9 committed:
remove the api coordinator

kclif9 committed:
use progress config flow to make authentication more seamless

kclif9 committed:
fix tests

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek reviewed (changes requested):
I think we're almost there, feel free to send me a message on Discord because I would love to get this over the finish line :)

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Merge branch 'home-assistant:dev' into add-actronairneo-climate-integration

kclif9 committed:
add exception for CannotConnect

kclif9 committed:
Set config flow to handle timeout properly

kclif9 committed:
Rename integration from actronair to actron_air

kclif9 committed:
Remove other references to Neo where feasible

kclif9 committed:
Refactor Actron Air config flow tests to use ActronAirConfigFlow and add mock_actron_api fixture

kclif9 committed:
Enhance Actron Air config flow tests with improved mock API and async handling

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek reviewed (changes requested):
I think this PR is almost there, be sure to message me on Discord :)

kclif9 committed:
Update homeassistant/components/actron_air/climate.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update homeassistant/components/actron_air/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update homeassistant/components/actron_air/coordinator.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update tests/components/actron_air/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update tests/components/actron_air/conftest.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update tests/components/actron_air/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update tests/components/actron_air/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Update tests/components/actron_air/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

kclif9 committed:
Merge branch 'dev' into add-actronairneo-climate-integration

kclif9 committed:
Refactor OAuth2 flow tests for improved clarity and add duplicate account handling

joostlek committed:
Apply suggestions from code review

joostlek committed:
Update tests/components/actron_air/test_config_flow.py

joostlek committed:
Update tests/components/actron_air/test_config_flow.py

joostlek committed:
Merge branch 'dev' into add-actronairneo-climate-integration

joostlek committed:
Fix
